xbmc/interfaces/python/PythonInvoker.cpp

```diff
@@ -683,23 +678,22 @@ void CPythonInvoker::onError(const std::string& exceptionType /* = "" */,
   }
 }
 
-void CPythonInvoker::initializeModules(
-    const std::map<std::string, PythonModuleInitialization>& modules)
+void CPythonInvoker::initializeModules(const PythonModule modules[], const int numElements)
 {
-  for (const auto& module : modules)
+  for (int i = 0; i < numElements; i++)
   {
-    if (!initializeModule(module.second))
+    if (!initializeModule(modules[i].name, modules[i].initialization))
       CLog::Log(LOGWARNING, "CPythonInvoker({}, {}): unable to initialize python module \"{}\"",
-                GetId(), m_sourceFile, module.first);
+                GetId(), m_sourceFile, modules[i].name);
   }
 }
 
-bool CPythonInvoker::initializeModule(PythonModuleInitialization module)
+bool CPythonInvoker::initializeModule(const char* name, PythonModuleInitialization module)
 {
   if (module == NULL)
     return false;
 
-  return module() != nullptr;
+  return !PyImport_AppendInittab(name, module);
```

Comment:
While looking at #22344 I've noticed that inittab is a global resource in Python. Each module should only be added once. Modifying the inittab later may even [crash](http://ix.io/4kyb).

My current proof proof concept (without this PR) is:
```
From dc07e8b90821896b4d529fb6abb57d04e7fac88f Mon Sep 17 00:00:00 2001
From: mglae <mglmail@arcor.de>
Date: Thu, 5 Jan 2023 00:35:22 +0100
Subject: [PATCH] AddonPythonInvoker: Only call PyImport_AppendInittab() once

The modules are added to a single Inittab so
a) The tab will grow with any call and duplicate entries are added.
b) The tab check code may crash while the tab is expanded.
---
 xbmc/interfaces/python/AddonPythonInvoker.cpp | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/xbmc/interfaces/python/AddonPythonInvoker.cpp b/xbmc/interfaces/python/AddonPythonInvoker.cpp
index b6158a8db6..34280f48ac 100644
--- a/xbmc/interfaces/python/AddonPythonInvoker.cpp
+++ b/xbmc/interfaces/python/AddonPythonInvoker.cpp
@@ -103,12 +103,18 @@ static PythonModule PythonModules[] =
 CAddonPythonInvoker::CAddonPythonInvoker(ILanguageInvocationHandler *invocationHandler)
   : CPythonInvoker(invocationHandler)
 {
-  PyImport_AppendInittab("xbmcdrm", PyInit_Module_xbmcdrm);
-  PyImport_AppendInittab("xbmcgui", PyInit_Module_xbmcgui);
-  PyImport_AppendInittab("xbmc", PyInit_Module_xbmc);
-  PyImport_AppendInittab("xbmcplugin", PyInit_Module_xbmcplugin);
-  PyImport_AppendInittab("xbmcaddon", PyInit_Module_xbmcaddon);
-  PyImport_AppendInittab("xbmcvfs", PyInit_Module_xbmcvfs);
+  static bool first_init(true);
+
+  if (first_init)
+  {
+    first_init = false;
+    PyImport_AppendInittab("xbmcdrm", PyInit_Module_xbmcdrm);
+    PyImport_AppendInittab("xbmcgui", PyInit_Module_xbmcgui);
+    PyImport_AppendInittab("xbmc", PyInit_Module_xbmc);
+    PyImport_AppendInittab("xbmcplugin", PyInit_Module_xbmcplugin);
+    PyImport_AppendInittab("xbmcaddon", PyInit_Module_xbmcaddon);
+    PyImport_AppendInittab("xbmcvfs", PyInit_Module_xbmcvfs);
+  }
 }
 
 CAddonPythonInvoker::~CAddonPythonInvoker() = default;
-- 
2.35.3
```

Reply:
The current code seems to implement a kind of "sub interpreter". This feature is not properly documented, and requires the function PyImport_AppendInittab() to be called locally. This new implementation doesn't change the previous behavior.

Here is an implementation of a ""sub interpreter": https://gist.github.com/sterin/61561c3139dd49da1f43

Reply:
> The current code seems to implement a kind of "sub interpreter". This feature is not properly documented, and requires the function PyImport_AppendInittab() to be called locally. 

PyImport_AppendInittab()  modifies a global table and should be called before Py_Initialize(), see comment and implementation [here](https://github.com/python/cpython/blob/v3.11.1/Python/import.c#L2570-L2634).

Reply:
The https://github.com/xbmc/xbmc/issues/22344 issue seems to be related to thread safety, I doubt changing  PyImport_AppendInittab() would fix anything, but the gcc sanitizer could help to find where the issue is.

Reply:
#22344 is a Python regression and will be fixed in the next 3.10/3.11 dot releases.

But while testing I noticed the PyImport_AppendInittab() related crashes.

The following patch to this PR is working fine in my Linux builds with external Python 3.6.15 and 3.11.1:
```
diff --git a/xbmc/interfaces/python/AddonPythonInvoker.cpp b/xbmc/interfaces/python/AddonPythonInvoker.cpp
index ca6b732af3..ce36309513 100644
--- a/xbmc/interfaces/python/AddonPythonInvoker.cpp
+++ b/xbmc/interfaces/python/AddonPythonInvoker.cpp
@@ -103,7 +103,13 @@ const std::array<PythonModule, 6> PythonModules =
 CAddonPythonInvoker::CAddonPythonInvoker(ILanguageInvocationHandler *invocationHandler)
   : CPythonInvoker(invocationHandler)
 {
-  initializeModules(PythonModules);
+  static bool first_init(true);
+
+  if (first_init)
+  {
+    first_init = false;
+    initializeModules(PythonModules);
+  }
 }
 
 CAddonPythonInvoker::~CAddonPythonInvoker() = default;
```

Reply:
@mglae, your patch is working, but we should keep in mind the official python documentation PyImport_AppendInittab(): - "This should be called before Py_Initialize()" https://docs.python.org/3/c-api/import.html

Here is a patch compliant with the python documentation:
```
diff --git a/xbmc/interfaces/python/AddonPythonInvoker.cpp b/xbmc/interfaces/python/AddonPythonInvoker.cpp
index ca6b732af3..b360c7d45d 100644
--- a/xbmc/interfaces/python/AddonPythonInvoker.cpp
+++ b/xbmc/interfaces/python/AddonPythonInvoker.cpp
@@ -103,11 +103,15 @@ const std::array<PythonModule, 6> PythonModules =
 CAddonPythonInvoker::CAddonPythonInvoker(ILanguageInvocationHandler *invocationHandler)
   : CPythonInvoker(invocationHandler)
 {
-  initializeModules(PythonModules);
 }
 
 CAddonPythonInvoker::~CAddonPythonInvoker() = default;
 
+void CAddonPythonInvoker::globalInitializeModules(void)
+{
+  initializeModules(PythonModules);
+}
+
 const char* CAddonPythonInvoker::getInitializationScript() const
 {
   return RUNSCRIPT_COMPLIANT;
diff --git a/xbmc/interfaces/python/AddonPythonInvoker.h b/xbmc/interfaces/python/AddonPythonInvoker.h
index 5ba0786438..cacf144ddd 100644
--- a/xbmc/interfaces/python/AddonPythonInvoker.h
+++ b/xbmc/interfaces/python/AddonPythonInvoker.h
@@ -16,6 +16,8 @@ public:
   explicit CAddonPythonInvoker(ILanguageInvocationHandler *invocationHandler);
   ~CAddonPythonInvoker() override;
 
+  static void globalInitializeModules(void);
+
 protected:
   // overrides of CPythonInvoker
   const char* getInitializationScript() const override;
diff --git a/xbmc/interfaces/python/PythonInvoker.h b/xbmc/interfaces/python/PythonInvoker.h
index 4583e1629a..09b68c2f0e 100644
--- a/xbmc/interfaces/python/PythonInvoker.h
+++ b/xbmc/interfaces/python/PythonInvoker.h
@@ -60,18 +60,18 @@ protected:
 
 protected:
   template<size_t NUMBER>
-  void initializeModules(const std::array<PythonModule, NUMBER> modules)
+  static void initializeModules(const std::array<PythonModule, NUMBER> modules)
   {
     for (const auto& module : modules)
     {
       if (!initializeModule(module.name, module.initialization))
-        CLog::Log(LOGWARNING, "CPythonInvoker({}, {}): unable to initialize python module \"{}\"",
-                  GetId(), m_sourceFile, module.name);
+        CLog::Log(LOGWARNING, "CPythonInvoker(): unable to initialize python module \"{}\"",
+                  module.name);
     }
   }
 
 private:
-  bool initializeModule(const char* name, PythonModuleInitialization module);
+  static bool initializeModule(const char* name, PythonModuleInitialization module);
   void getAddonModuleDeps(const ADDON::AddonPtr& addon, std::set<std::string>& paths);
   bool execute(const std::string& script, std::vector<std::wstring>& arguments);
   FILE* PyFile_AsFileWithMode(PyObject* py_file, const char* mode);
diff --git a/xbmc/interfaces/python/XBPython.cpp b/xbmc/interfaces/python/XBPython.cpp
index 9cb9ba00ac..86079d1d28 100644
--- a/xbmc/interfaces/python/XBPython.cpp
+++ b/xbmc/interfaces/python/XBPython.cpp
@@ -34,6 +34,10 @@
 #include "platform/Environment.h"
 #endif
 
+#ifdef HAS_WEB_INTERFACE
+#include "network/httprequesthandler/python/HTTPPythonWsgiInvoker.h"
+#endif
+
 #include <algorithm>
 
 // Only required for Py3 < 3.7
@@ -526,6 +530,12 @@ bool XBPython::OnScriptInitialized(ILanguageInvoker* invoker)
     Py_OptimizeFlag = 1;
 #endif
 
+    CAddonPythonInvoker::globalInitializeModules();
+
+#ifdef HAS_WEB_INTERFACE
+    CHTTPPythonWsgiInvoker::globalInitializeModules();
+#endif
+
     Py_Initialize();
 
 #if PY_VERSION_HEX < 0x03070000
diff --git a/xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp b/xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp
index 3bc46c1b76..1dd3f2d33a 100644
--- a/xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp
+++ b/xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp
@@ -92,7 +92,6 @@ CHTTPPythonWsgiInvoker::CHTTPPythonWsgiInvoker(ILanguageInvocationHandler* invoc
   : CHTTPPythonInvoker(invocationHandler, request),
     m_wsgiResponse(NULL)
 {
-  initializeModules(PythonModules);
 }
 
 CHTTPPythonWsgiInvoker::~CHTTPPythonWsgiInvoker()
@@ -101,6 +100,11 @@ CHTTPPythonWsgiInvoker::~CHTTPPythonWsgiInvoker()
   m_wsgiResponse = NULL;
 }
 
+void CHTTPPythonWsgiInvoker::globalInitializeModules(void)
+{
+  initializeModules(PythonModules);
+}
+
 HTTPPythonRequest* CHTTPPythonWsgiInvoker::GetRequest()
 {
   if (m_request == NULL || m_wsgiResponse == NULL)
diff --git a/xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.h b/xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.h
index 73f9ccf0be..86f075e517 100644
--- a/xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.h
+++ b/xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.h
@@ -29,6 +29,8 @@ public:
   CHTTPPythonWsgiInvoker(ILanguageInvocationHandler* invocationHandler, HTTPPythonRequest* request);
   ~CHTTPPythonWsgiInvoker() override;
 
+  static void globalInitializeModules(void);
+
   // implementations of CHTTPPythonInvoker
   HTTPPythonRequest* GetRequest() override;
 
```

Reply:
@repojohnray Agree, this is the correct way and it's working fine.

I can confirm this commit fixes the memory leaks.

Reply:
@mglae, I will update the commit.